[GTK][WPE] Use WidthIterator for simple text when Font::applyTransforms is not needed

carlosgcampos opened the pull request:
#### 8cb04a3ecae37624cfba82f6e6668e020a7eb477
<pre>
[GTK][WPE] Use WidthIterator for simple text when Font::applyTransforms is not needed
<a href="https://bugs.webkit.org/show_bug.cgi?id=299278">https://bugs.webkit.org/show_bug.cgi?id=299278</a>

Reviewed by Fujii Hironori.

In 299354@main we introduced a build option to always use
ComplexTextController even for simple text code path. However, there are
cases in which we can use WidthIterator instead, for example when
optimize for speed rendering mode is used, because we know
Font::ApplyTransforms can be skipped. This patch removes that build
option, to make it possible to make the decision at runtime.

Canonical link: <a href="https://commits.webkit.org/300567@main">https://commits.webkit.org/300567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7428c9f32dc9b8bbd0b2ffa86ec9afe71f0bd002

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123104 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129765 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124981 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43542 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51413 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93578 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126055 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110176 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74212 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28332 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73278 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115263 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104416 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28558 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132492 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121635 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50054 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38107 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102076 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50430 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106397 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101928 "Found 3 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25899 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47303 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25500 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49909 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55670 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151900 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49377 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38839 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52729 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51058 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -631,6 +653,31 @@ bool FontCascade::shouldDisableFontSubpixelAntialiasingForTesting()
     return disableFontSubpixelAntialiasingForTesting;
 }
 
+bool FontCascade::canHandleRunWithSimpleText(const TextRun& run, unsigned from, unsigned to) const
```
canHandleRunWithSimpleText isn't a good name.
Complext text is a text requiring shaping, for example, Arabic, Hebrew and Emoji.
Simple texts is a text requiring only kerning and ligature, for example, English.
It's a problem of WidthIterator. https://github.com/WebKit/WebKit/blob/6a265e4581b6fffff53c4b35782ce2ed6b306f37/Source/WebCore/platform/graphics/FontCascade.cpp#L1447-L1448
What about canHandleRunWithWidthIterator or canHandleRunWithSimpleTextCodePath?

carlosgcampos replied:
I disagree, this is used to decide the FontCascade::CodePath, then shouldUseComplexTextController decides whether to use WidthIterator or ComplexTextController. I agree this is because of a limitation of WidthIterator, but it's still deciding the CodePath at that point.

fujii replied:
Then, how about canHandleRunAsSimpleText?

carlosgcampos replied:
Sounds good to me too

carlosgcampos committed:
[GTK][WPE] Use WidthIterator for simple text when Font::applyTransforms is not needed
https://bugs.webkit.org/show_bug.cgi?id=299278

Reviewed by Fujii Hironori.

In 299354@main we introduced a build option to always use
ComplexTextController even for simple text code path. However, there are
cases in which we can use WidthIterator instead, for example when
optimize for speed rendering mode is used, because we know
Font::ApplyTransforms can be skipped. This patch removes that build
option, to make it possible to make the decision at runtime.

Canonical link: https://commits.webkit.org/300567@main
